pretyping/classops.ml

```diff
@@ -292,12 +292,19 @@ let lookup_path_to_fun_from env sigma s =
 let lookup_path_to_sort_from env sigma s =
   apply_on_class_of env sigma s lookup_path_to_sort_from_class
 
+let mkNamed = function
+  | GlobRef.ConstRef c -> EConstr.mkConst c
+  | VarRef v -> EConstr.mkVar v
+  | ConstructRef c -> EConstr.mkConstruct c
+  | IndRef i -> EConstr.mkInd i
+
 let get_coercion_constructor env coe =
-  match coe.coe_value with
-  | ConstructRef c ->
-      (c, Inductiveops.constructor_nrealargs c -1)
-  | _ ->
-      raise Not_found
+  let evd = Evd.from_env env in
+  let red x = fst (Reductionops.whd_all_stack env evd x) in
```

Comment:
This is using the reductionops reduction that is very inefficient. Because you're just looking at the head constructor of an expression, you should use `Reductionops.clos_whd_flags` instead.

Reply:
Unless I'm mistaken, I used the function that was used before.

Reply:
That doesn't make it right, unfortunately. Calls to the Reductionops machine should be really considered twice, they're rarely doing what you're actually wanting.

Reply:
> That doesn't make it right, unfortunately. Calls to the Reductionops machine should be really considered twice, they're rarely doing what you're actually wanting.

Is the message "you should should call the `Reductionops` machine, but instead use `Reductionops.clos_whd_flags`"? That sounds like a joke.

Reply:
`Reductionops.clos_whd_flags` is not the Reductionops machine, it's the kernel machine. Maybe the phrasing *Reductionops machine* is misleading, but this module actually implements a full-blown (again, technically two slightly distinct with little code sharing) reduction machine that does wonders like refolding, russian motorbikes and *retour de l'être aimé*.

Reply:
This is a change that is completely orthogonal to this PR. I'll be willing to do it, even on a large scale (because many pieces are calling the wrong one, I think), but only once we have an understandable API.

Reply:
Addendum: `clos` probably stands for `(C)Closure`.